src/lib/support/Pool.h

```diff
@@ -160,4 +167,107 @@ class BitMapObjectPool : public StaticAllocatorBitmap
     } mData;
 };
 
+#if CHIP_SYSTEM_CONFIG_POOL_USE_HEAP
+
+/**
+ *  @brief
+ *      A class template used for allocating Object subclass objects from an ObjectArena<> template union.
+ *
+ *  @tparam     T   a class to be allocated from the arena.
+ */
+template <class T>
+class HeapObjectPool
+{
+public:
+    template <typename... Args>
+    T * CreateObject(Args &&... args)
+    {
+        std::lock_guard<std::mutex> lock(mutex);
+        T * object = new T(std::forward<Args>(args)...);
+        if (object == nullptr)
+        {
+            return nullptr;
+        }
+
+        mObjects.insert(object);
+        return object;
+    }
+
+    void ReleaseObject(T * object)
+    {
+        if (object == nullptr)
+            return;
+
+        std::lock_guard<std::mutex> lock(mutex);
+        auto iter = mObjects.find(object);
+        VerifyOrDie(iter != mObjects.end());
+        mObjects.erase(iter);
+        delete object;
+    }
+
+    template <typename... Args>
+    void ResetObject(T * element, Args &&... args)
+    {
+        element->~T();
+        new (element) T(std::forward<Args>(args)...);
+    }
+
+    /**
+     * @brief
+     *   Run a functor for each active object in the pool
+     *
+     *  @param     function The functor of type `bool (*)(T*)`, return false to break the iteration
+     *  @return    bool     Returns false if broke during iteration
+     */
+    template <typename Function>
+    bool ForEachActiveObject(Function && function)
+    {
+        std::lock_guard<std::mutex> lock(mutex);
+        // Create a new copy of original set, allowing add/remove elements while iterating in the same thread.
```

Comment:
I don't see how this comment makes sense.  `std::mutex` is not a recursive mutex, so trying to "add/remove elements while iterating in the same thread" will deadlock, no?

Reply:
Yes, I didn't revise this properly. There doesn't need to be a lock but it does need to allow the pool to change.

Reply:
Two things:

1. There needs to be a lock while making the copy, right?
2. Once you make the copy, a Release() of an object will cause the copy to hold a dangling reference....

I don't see a great way to address #2, honestly, while allowing mutation of the pool during the iteration.  :(

Reply:
It also appears the case of in-place object deletion with recursive ForEachNode calls is now handled.

Thanks for that.